src/main/java/org/opensearch/cluster/etcd/changeapplier/DataNodeState.java

```diff
@@ -140,15 +215,59 @@ public ClusterState buildClusterState(ClusterState previousState) {
                 ShardRouting shardRouting;
                 if (previouslyStartedShard.isPresent()) {
                     shardRouting = previouslyStartedShard.get();
+                    logger.debug(
+                        "Reusing existing ShardRouting for shard {}[{}] with allocation ID {}",
+                        indexMetadata.getIndex().getName(),
+                        shardNum,
+                        shardRouting.allocationId().getId()
+                    );
                 } else {
+                    // No previous shard in cluster state - use ETCD-based allocation ID preservation
+                    RecoverySource etcdRecoverySource = determineRecoverySource(shardNum, indexMetadata, role);
+                    String previousAllocationId = getPreviousAllocationId(indexMetadata.getIndex().getName(), shardNum);
+
                     shardRouting = ShardRouting.newUnassigned(
                         shardId,
                         role == ShardRole.PRIMARY,
                         role == ShardRole.SEARCH_REPLICA,
-                        recoverySource,
+                        etcdRecoverySource, // Use recovery source based on ETCD data
```

Comment:
I think this line is kind of a bad merge, since the `recoverySource` variable is now useless. 

At the very least, I think the logic to figure out which recovery source to use should move to around line 177. (Or we can remove the logic up there to figure out the recovery source and just do it here. Either works, but it should be one or the other.)